File: exercise/typescript/05_playlist_saver.ts
/**
 * Exercise 5: Playlist Saver
 *
 * Learn to work with JSON in TypeScript while maintaining type safety!
 * Since we're using browser-compatible TypeScript (no Node.js fs module),
 * we'll use JSON and localStorage concepts that work in browsers.
 *
 * Concepts covered:
 * - JSON.stringify() and JSON.parse()
 * - Type assertions with as keyword
 * - Working with localStorage (browser API)
 * - Type guards and validation
 * - Simulating data persistence
 *
 * Run this exercise: node exercise/typescript/05_playlist_saver.ts
 */

// ============================================================================
// ABOUT JSON AND TYPE SAFETY
// ============================================================================
// JSON (JavaScript Object Notation) is a text format for storing data.
// In TypeScript:
// - JSON.stringify() converts objects to JSON strings
// - JSON.parse() converts JSON strings back to objects
//
// IMPORTANT: JSON.parse() returns type 'any' (no type safety!)
// We use TYPE ASSERTIONS to tell TypeScript what type we expect.
//
// Note: This exercise simulates data persistence. In browsers, you'd use
// localStorage. In Node.js, you'd use the fs module. We're keeping it simple!


// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

interface Song {
    title: string;
    artist: string;
    duration: number;
    genre: string;
}

interface Playlist {
    name: string;
    songs: Song[];
    createdDate: string;
}


// ============================================================================
// EXAMPLE CODE
// ============================================================================

// Creating a sample playlist
const examplePlaylist: Playlist = {
    name: "Classic Rock Favorites",
    songs: [
        { title: "Imagine", artist: "John Lennon", duration: 3.03, genre: "Pop" },
        { title: "Billie Jean", artist: "Michael Jackson", duration: 4.54, genre: "Pop" },
        { title: "Bohemian Rhapsody", artist: "Queen", duration: 5.55, genre: "Rock" }
    ],
    createdDate: "2024-12-25"
};

console.log("🎵 Original Playlist 🎵");
console.log(`Playlist: ${examplePlaylist.name}`);
for (const song of examplePlaylist.songs) {
    console.log(`♪ ${song.title} by ${song.artist}`);
}


// Converting to JSON string
function playlistToJSON(playlist: Playlist): string {
    return JSON.stringify(playlist, null, 2);  // null, 2 for pretty formatting
}

const jsonString = playlistToJSON(examplePlaylist);
console.log("\n📄 JSON String:");
console.log(jsonString);


// Converting from JSON string back to object
function jsonToPlaylist(jsonString: string): Playlist {
    // JSON.parse returns 'any' type - we use 'as Playlist' to assert the type
    const playlist = JSON.parse(jsonString) as Playlist;
    return playlist;
}

const loadedPlaylist = jsonToPlaylist(jsonString);
console.log("\n🎵 Loaded from JSON 🎵");
console.log(`Playlist: ${loadedPlaylist.name}`);
console.log(`Songs: ${loadedPlaylist.songs.length}`);


// Simulating localStorage (browser API for storing data)
// Note: This won't actually persist in Node.js, but demonstrates the concept
const storage: { [key: string]: string } = {};  // Simulated storage

function savePlaylist(playlist: Playlist, key: string): void {
    const json = JSON.stringify(playlist);
    storage[key] = json;
    console.log(`\n✅ Saved playlist "${playlist.name}" to key: ${key}`);
}

function loadPlaylist(key: string): Playlist | null {
    const json = storage[key];
    if (!json) {
        console.log(`\n❌ No playlist found for key: ${key}`);
        return null;
    }
    const playlist = JSON.parse(json) as Playlist;
    console.log(`\n✅ Loaded playlist "${playlist.name}" from key: ${key}`);
    return playlist;
}

// Testing save and load
savePlaylist(examplePlaylist, "myPlaylist");
const retrieved = loadPlaylist("myPlaylist");
if (retrieved) {
    console.log(`Retrieved playlist has ${retrieved.songs.length} songs`);
}


// ============================================================================
// YOUR TURN: TODO EXERCISES
// ============================================================================

// TODO 1: Create your own playlist to save
// Create a Playlist object with at least 3 songs
// Make sure each song matches the Song interface!
// Stuck? Ask: "How do I create an object that matches an interface?"

const myPlaylist: Playlist = {
    // TODO: Fill in the playlist properties
    name: "",
    songs: [],
    createdDate: ""
};


// TODO 2: Convert your playlist to JSON
// Use JSON.stringify() to convert myPlaylist to a JSON string
// Print the JSON string to see what it looks like
// Meta-help: "What does JSON.stringify() do in TypeScript?"

// TODO: Create a variable jsonData and assign JSON.stringify(myPlaylist, null, 2)
// TODO: Print jsonData


// TODO 3: Parse the JSON back to an object
// Use JSON.parse() with a type assertion to convert the JSON back
// Type hint: const parsed = JSON.parse(jsonData) as Playlist;

// TODO: Parse jsonData back to a Playlist object
// TODO: Print the playlist name and song count to verify it worked


// TODO 4: Use the save and load functions
// Test the savePlaylist and loadPlaylist functions with your playlist

// TODO: Save myPlaylist with the key "myFavorites"
// TODO: Load it back using loadPlaylist("myFavorites")
// TODO: Print the loaded playlist's information


// TODO 5: Handle missing data
// Try to load a playlist that doesn't exist
// The function returns null - use an if statement to check

console.log("\n🔍 Testing Error Handling 🔍");
// TODO: Try to load a playlist with key "nonexistent"
// TODO: Check if the result is null
// TODO: Print an appropriate message


// TODO 6: Create a type guard function
// Write a function that checks if an object is a valid Song
// This is useful for validating JSON data!
// Meta-help: "What are type guards in TypeScript?"

function isSong(obj: any): obj is Song {
    // TODO: Check if obj has all required Song properties
    // TODO: Return true if valid, false otherwise
    // HINT: Check typeof obj.title === "string", etc.
    return false;  // Replace this
}

// Test your type guard
const testObj = { title: "Test", artist: "Artist", duration: 3.5, genre: "Rock" };
// TODO: Call isSong with testObj and print the result


// TODO 7: Create a function to add a song and save
// Write a function that:
// 1. Loads a playlist
// 2. Adds a new song
// 3. Saves it back
// This combines multiple concepts!

function addSongAndSave(playlistKey: string, newSong: Song): void {
    // TODO: Load the playlist using loadPlaylist
    // TODO: If playlist is null, print error and return
    // TODO: Add newSong to playlist.songs
    // TODO: Save the updated playlist
    // TODO: Print a confirmation message
}

// Test your function
const newSong: Song = {
    title: "Your Choice",
    artist: "Your Favorite Artist",
    duration: 3.30,
    genre: "Rock"
};
// TODO: Call addSongAndSave to add newSong to "myFavorites"


// TODO 8: Create multiple playlists
// Create and save at least 3 different themed playlists
// Examples: workout, chill, party, studying

// TODO: Create a workout playlist
// TODO: Create a chill playlist
// TODO: Create another themed playlist
// TODO: Save each one with a different key
// TODO: Load them back and display their names


// TODO 9: Create a function to list all saved playlists
// Write a function that returns all playlist keys from storage
// Type hint: Object.keys(storage) returns string[]

function listAllPlaylists(): string[] {
    // TODO: Return all keys from the storage object
    return [];  // Replace this
}

// TODO: Call listAllPlaylists and print the results


// TODO 10: Create a function to merge playlists
// Write a function that takes two playlist keys and creates a combined playlist
// Remove duplicate songs (same title and artist)

function mergePlaylists(key1: string, key2: string, newName: string): Playlist | null {
    // TODO: Load both playlists
    // TODO: If either is null, return null
    // TODO: Combine the songs arrays
    // TODO: Remove duplicates (this is challenging!)
    // TODO: Create and return a new Playlist object
    return null;  // Replace this
}

// Test your merge function
// TODO: Create two playlists, save them, then merge them


// ============================================================================
// BONUS CHALLENGES (Optional)
// ============================================================================

// BONUS 1: Add data validation
// Create a function validatePlaylist(obj: any): obj is Playlist
// that checks if an object is a valid Playlist with valid Songs
// Use your isSong function to validate each song!


// BONUS 2: Create a backup system
// Write a function that saves a playlist to two keys: the original and a backup
// Example: saveWithBackup saves to "myPlaylist" and "myPlaylist_backup"


// BONUS 3: Add playlist metadata
// Extend the Playlist interface to include:
// - totalDuration (calculated from songs)
// - lastModified (timestamp)
// Create functions to automatically update these when saving


// BONUS 4: Implement search functionality
// Create a function searchAllPlaylists(query: string): Song[]
// that searches through ALL saved playlists for songs matching the query


// ============================================================================
// TESTING YOUR CODE
// ============================================================================

// When you run this file, you should see:
// - Example playlist converted to JSON and back
// - Your playlist saved and loaded
// - Error handling when loading nonexistent playlists
// - Type guard validating song objects
// - Songs being added to saved playlists
// - Multiple themed playlists created and saved
// - List of all saved playlist keys
// - Playlists being merged together

// Run with: node exercise/typescript/05_playlist_saver.ts

export {};

File: exercise/python/05_playlist_saver.py
"""
Exercise 5: Playlist Saver

Learn how to save your playlists to files and load them back later! This exercise
teaches file I/O (input/output) and JSON format - skills used in real applications.

Concepts covered:
- File I/O (reading and writing files)
- JSON format (JavaScript Object Notation)
- Error handling with try/except
- Data persistence
- Working with file paths

Learning Goals:
- Save data to files so it persists between program runs
- Read data from files
- Use JSON to store complex data structures
- Handle errors gracefully when files don't exist

Run this exercise: uv run exercise/python/05_playlist_saver.py
"""

import json
import os

# ============================================================================
# WHAT IS FILE I/O AND JSON?
# ============================================================================
# File I/O lets you save data to your computer and load it later.
# JSON is a text format for storing structured data (like dictionaries and lists).
#
# Why JSON?
# - Easy to read and write
# - Works with dictionaries, lists, strings, numbers
# - Used everywhere: web APIs, config files, data storage
#
# Python's json module converts Python data to JSON and back:
# - json.dump() / json.dumps() -> Python to JSON
# - json.load() / json.loads() -> JSON to Python


# ============================================================================
# EXAMPLE CODE
# ============================================================================

# Creating a sample playlist (list of song dictionaries)
example_playlist = [
    {"title": "Imagine", "artist": "John Lennon", "duration": 3.03},
    {"title": "Billie Jean", "artist": "Michael Jackson", "duration": 4.54},
    {"title": "Bohemian Rhapsody", "artist": "Queen", "duration": 5.55}
]

print("🎵 Original Playlist 🎵")
for song in example_playlist:
    print(f"♪ {song['title']} by {song['artist']}")


# Saving the playlist to a file
def save_playlist(playlist, filename):
    """Save a playlist to a JSON file."""
    with open(filename, 'w') as file:
        json.dump(playlist, file, indent=2)  # indent=2 makes it pretty
    print(f"\n✅ Playlist saved to {filename}")


# Loading a playlist from a file
def load_playlist(filename):
    """Load a playlist from a JSON file."""
    try:
        with open(filename, 'r') as file:
            playlist = json.load(file)
        print(f"\n✅ Playlist loaded from {filename}")
        return playlist
    except FileNotFoundError:
        print(f"\n❌ File {filename} not found!")
        return []


# Testing save and load
save_playlist(example_playlist, "example_playlist.json")
loaded_playlist = load_playlist("example_playlist.json")

print("\n🎵 Loaded Playlist 🎵")
for song in loaded_playlist:
    print(f"♪ {song['title']} by {song['artist']}")


# ============================================================================
# YOUR TURN: TODO EXERCISES
# ============================================================================

# TODO 1: Create your own playlist to save
# Make a list of at least 3 song dictionaries with title, artist, duration
# HINT: Use what you learned in Exercise 4

my_playlist = []  # TODO: Add your song dictionaries here


# TODO 2: Save your playlist to a file
# Call the save_playlist function to save my_playlist to "my_playlist.json"
# Stuck? Ask: "How do I call a function that takes two parameters?"

# TODO: Call save_playlist with my_playlist and "my_playlist.json"


# TODO 3: Load your playlist back
# Call load_playlist to load "my_playlist.json" and store it in a variable
# Print the loaded playlist to verify it worked
# Meta-help: "How do I load data from a JSON file in Python?"

# TODO: Call load_playlist and store the result in loaded_my_playlist
# TODO: Print the loaded playlist (loop through and display songs)


# TODO 4: Add error handling
# Try to load a file that doesn't exist and see what happens
# The load_playlist function already has error handling - test it!

print("\n🔍 Testing Error Handling 🔍")
# TODO: Try to load a file that doesn't exist, like "nonexistent.json"
# TODO: Notice how the try/except prevents the program from crashing


# TODO 5: Create a function to add a song and auto-save
# Write a function that:
# 1. Loads the playlist from a file
# 2. Adds a new song to it
# 3. Saves it back to the file
# This is a practical pattern used in real apps!

def add_song_and_save(filename, new_song):
    """Add a song to a playlist file and save it."""
    # TODO: Load the playlist from filename
    # TODO: Append new_song to the playlist
    # TODO: Save the updated playlist back to filename
    # TODO: Print a confirmation message
    pass  # Remove this when you add your code


# Test your function
new_song = {"title": "Your Choice", "artist": "Your Favorite Artist", "duration": 3.30}
# TODO: Call add_song_and_save to add new_song to "my_playlist.json"
# TODO: Load and print the playlist to verify the song was added


# TODO 6: Create a function to remove a song by title
# Write a function that loads a playlist, removes a song by title, and saves it back
# HINT: Use a loop to find the song, then use list.remove()

def remove_song_and_save(filename, song_title):
    """Remove a song from a playlist file by title."""
    # TODO: Load the playlist
    # TODO: Loop through to find the song with matching title
    # TODO: Remove it from the list
    # TODO: Save the updated playlist
    # TODO: Print a confirmation
    pass  # Remove this when you add your code


# Test it
# TODO: Call remove_song_and_save to remove a song from "my_playlist.json"


# TODO 7: Create multiple playlists
# Create and save at least 3 different playlists with different themes
# Examples: "workout.json", "chill.json", "party.json"
# Meta-help: "How can I organize multiple playlists in my music app?"

# TODO: Create a workout playlist (energetic songs)
# TODO: Create a chill playlist (relaxing songs)
# TODO: Create another themed playlist of your choice
# TODO: Save each one to a different file


# TODO 8: Create a playlist manager menu
# This combines everything! Create a simple text-based menu that lets you:
# - List all saved playlists
# - Load and display a playlist
# - Create a new playlist
# - Add songs to a playlist
# - Save playlists
# This is advanced but rewarding!

def playlist_menu():
    """A simple playlist manager with a menu."""
    print("\n🎵 Playlist Manager 🎵")
    print("1. List all playlists")
    print("2. Load a playlist")
    print("3. Create new playlist")
    print("4. Add song to playlist")
    print("5. Save and exit")

    # TODO: Get user input with input()
    # TODO: Use if/elif statements to handle each menu option
    # TODO: This is a big challenge - break it into small steps!
    # Meta-help: "How do I create a menu-driven program in Python?"
    pass


# Uncomment to test your menu:
# playlist_menu()


# ============================================================================
# BONUS CHALLENGES (Optional)
# ============================================================================

# BONUS 1: Add data validation
# Before saving, check that each song has required fields (title, artist, duration)
# Print warnings for invalid songs
# HINT: Use "key" in dictionary to check if a key exists


# BONUS 2: Create a backup function
# Write a function that saves a playlist to both .json and a backup file
# Example: save_with_backup("my_playlist.json") also creates "my_playlist.backup.json"


# BONUS 3: Merge playlists
# Create a function that loads two playlists and combines them into one
# Remove duplicates (same title and artist)
# Meta-help: "How do I merge two lists and remove duplicates in Python?"


# BONUS 4: Export playlist as text
# Create a function that saves a playlist as a human-readable .txt file
# Format: "1. Song Title by Artist (3:30)"
# HINT: You'll need to convert decimal duration to minutes:seconds


# ============================================================================
# TESTING YOUR CODE
# ============================================================================

# When you run this file, you should see:
# - Example playlist saved and loaded
# - Your playlist saved to my_playlist.json
# - Your playlist loaded back successfully
# - Error handling working when loading nonexistent files
# - Songs being added and removed from the saved file
# - Multiple themed playlists created and saved

# Check the exercise/python/ folder - you should see JSON files created!
